gapir/cc/interpreter.cpp

```diff
@@ -302,46 +582,104 @@ Interpreter::Result Interpreter::load(uint32_t opcode) {
     GAPID_WARNING("Error: load not readable address %p", address);
     return ERROR;
   }
-  mStack.pushFrom(type, address);
-  return mStack.isValid() ? SUCCESS : ERROR;
+
+  switch (type) {
+    case BaseType::Bool:
+      mStack.push(*(bool*)address);
+      break;
+
+    case BaseType::Int8:
+      mStack.push(*(int8_t*)address);
+      break;
+
+    case BaseType::Int16:
+      mStack.push(*(int16_t*)address);
+      break;
+
+    case BaseType::Int32:
+      mStack.push(*(int32_t*)address);
+      break;
+
+    case BaseType::Int64:
+      mStack.push(*(int64_t*)address);
+      break;
+
+    case BaseType::Uint8:
+      mStack.push(*(uint8_t*)address);
+      break;
+
+    case BaseType::Uint16:
+      mStack.push(*(uint16_t*)address);
+      break;
+
+    case BaseType::Uint32:
+      mStack.push(*(uint32_t*)address);
+      break;
+
+    case BaseType::Uint64:
+      mStack.push(*(uint64_t*)address);
+      break;
+
+    case BaseType::Float:
+      mStack.push(*(float*)address);
+      break;
+
+    case BaseType::Double:
+      mStack.push(*(double*)address);
+      break;
+
+    case BaseType::AbsolutePointer:
+      mStack.push(*(void**)address);
+      break;
+
+    case BaseType::ConstantPointer:
+      mStack.push(Stack::ConstantPointer(*(uint32_t*)address));
+      break;
+
+    case BaseType::VolatilePointer:
+      mStack.push(Stack::VolatilePointer(*(uint32_t*)address));
+      break;
+
+    default:
+      assert(false);  // Unhandled type in loadC
+      break;
+  }
+
+  return SUCCESS;
 }
 
 Interpreter::Result Interpreter::pop(uint32_t opcode) {
   mStack.discard(extract26bitData(opcode));
-  return mStack.isValid() ? SUCCESS : ERROR;
+  return SUCCESS;
 }
 
 Interpreter::Result Interpreter::storeV(uint32_t opcode) {
   void* address = mMemoryManager->volatileToAbsolute(extract26bitData(opcode));
-  if (!isVolatileAddressForType(address, mStack.getTopType())) {
-    GAPID_WARNING("Error: storeV not volatile address %p", address);
-    return ERROR;
-  }
-
   mStack.popTo(address);
-  return mStack.isValid() ? SUCCESS : ERROR;
+  return SUCCESS;
 }
 
-Interpreter::Result Interpreter::store() {
+Interpreter::Result Interpreter::store(uint32_t) {
   void* address = mStack.pop<void*>();
   if (!isWriteAddress(address)) {
     GAPID_WARNING("Error: store not write address %p", address);
     return ERROR;
   }
   mStack.popTo(address);
-  return mStack.isValid() ? SUCCESS : ERROR;
+  return SUCCESS;
 }
 
 Interpreter::Result Interpreter::resource(uint32_t opcode) {
-  mStack.push<uint32_t>(extract26bitData(opcode));
+  auto id = extract26bitData(opcode);
+  mStack.push<uint32_t>(id);
   return this->call(Interpreter::RESOURCE_FUNCTION_ID);
 }
 
-Interpreter::Result Interpreter::post() {
+Interpreter::Result Interpreter::post(uint32_t) {
```

Comment:
Why do these take uint32_t values now?

Reply:
For symmetry with the other functions. I was playing with replacing the switch function with an array of function pointers and it was required then to get the same signature on each function. I can remove the unused int if you prefer it.

Reply:
Might be good to at least name the parameter to prevent confusion?

Reply:
Some people leave unused arguments nameless to make it clear they're not used. I picked this up from past software I worked on. TBH I prefer always naming. Changed.